readthedocs/core/views/hooks.py

```diff
@@ -42,16 +42,26 @@ def _build_version(project, slug, already_built=()):
         # these will build at "latest", and thus won't be
         # active
         latest_version = project.versions.get(slug=LATEST)
-        trigger_build(project=project, version=latest_version, force=True)
-        log.info("(Version build) Building %s:%s",
-                 project.slug, latest_version.slug)
-        if project.versions.exclude(active=False).filter(slug=slug).exists():
-            # Handle the case where we want to build the custom branch too
-            slug_version = project.versions.get(slug=slug)
-            trigger_build(project=project, version=slug_version, force=True)
-            log.info("(Version build) Building %s:%s",
-                     project.slug, slug_version.slug)
-        return LATEST
+        if latest_version.active:
+            trigger_build(project=project, version=latest_version, force=True)
+            log.info(
+                "(Version build) Building %s:%s",
+                project.slug, latest_version.slug
+            )
+            version_exists = (
+                project.versions.exclude(active=False)
+                .filter(slug=slug)
+                .exists()
+            )
+            if version_exists:
```

Comment:
Is there a reason we want to do this check only when the latest version is active? I'm not 100% sure, but it seems like we'd want it to run regardless of the state of the `latest` version.

Reply:
A user deactivates latest, here we try to build latest when the branch is `master`, as a side effect the version gets activated.

I just realized that we may need this strange logic, as this allows us to update the version list... Maybe we want to merge this after https://github.com/rtfd/readthedocs.org/issues/4450 ?

Reply:
I'm not sure to follow @stsewd. If the branch is `master` we will have two versions: master and latest.

When a webhook for `master` is received, we will want to build both if they are enabled. If `latest` is **disabled** we will want to build `master` version. Right?

Reply:
Yup, `master` is still building here, the code above handle that case, that's why I'm suggesting to remove this block too, as I think it only adds complexity to the code. But without building `latest` (activated or not) we don't have a way of updating the branch list, users will need to trigger a build to any other version to get the latest tags/branches on rtd.

Reply:
> But without building `latest` (activated or not) we don't have a way of updating the branch list, users will need to trigger a build to any other version to get the latest tags/branches on rtd.

A similar case is already happening in RTD:

1. import your project
1. create a new branch in your repository
1. go to Versions tab in RTD

Your new branch is not there.

I think we should implement a completely separate way of syncing our version with the ones on the repo. I mean, we should not build _anything_ if we receive a webhook, but update the versions list first. Then, decide what we need to build and trigger build those versions.

(this is probably out of the scope of this PR, but maybe worth thinking a little about this here)

> that's why I'm suggesting to remove this block too, as I think it only adds complexity to the code

This is the part where I don't follow you :(

Reply:
https://github.com/rtfd/readthedocs.org/issues/4450 this issue is trying to solve the sync problem, so, maybe is worth to solve that first? What do you think?

> This is the part where I don't follow you :(

This line just check if we are receiving a _main_ branch

https://github.com/stsewd/readthedocs.org/blob/76da8e5330c0523e6acc61ea561b3fb0b733e3d4/readthedocs/core/views/hooks.py#L40-L40

That way we build latest and the main branch at the same time. For the rest of versions, this is the normal flow

https://github.com/stsewd/readthedocs.org/blob/76da8e5330c0523e6acc61ea561b3fb0b733e3d4/readthedocs/core/views/hooks.py#L70-L79

So, at the end we are just saving one loop, even we aren't saving queries to the db now

Reply:
> #4450 this issue is trying to solve the sync problem, so, maybe is worth to solve that first? What do you think?

I'm fine of merging this PR first if it doesn't introduce another issue/bug/wrong workflow/confuse users.

#4450 is already a problem, and this PR doesn't fix it. Which is fine to me. I don't want to fix that issue in this PR, _but_ I want this statement to be correct to merge:

> When a webhook for master is received, we will want to build both if they are enabled. If latest is disabled we will want to build master version.

Anyway, I'm still confused with this chunk of code and it seems that talking async doesn't clarify things. We should chat/talk in sync mode.